karapace/backup/api.py

```diff
@@ -338,109 +334,115 @@ def __init__(
+        )
+
+        with _consumer(self.config, self.topic_name) as consumer:
             (partition,) = consumer.partitions_for_topic(self.topic_name)
             topic_partition = TopicPartition(self.topic_name, partition)
-            start_offset: int = consumer.beginning_offsets([topic_partition])[topic_partition]
-            end_offset: int = consumer.end_offsets([topic_partition])[topic_partition]
-            last_offset = start_offset
-            record_count = 0
-
-            fp.write(version.marker)
-            if start_offset < end_offset:  # non-empty topic
-                end_offset -= 1  # high watermark to actual end offset
-                print(
-                    (
-                        f"Started backup in format {version.name} of "
-                        f"{topic}:{partition} "
-                        f"(offset {start_offset:,} to {end_offset:,}) ..."
-                    ),
-                    file=sys.stderr,
+
+            try:
+                data_file = _write_partition(
+                    path=self.backup_location,
+                    backend=backend,
+                    consumer=consumer,
+                    topic_partition=topic_partition,
+                    poll_timeout=poll_timeout,
+                    allow_overwrite=overwrite,
                 )
-                while True:
-                    records: Collection[ConsumerRecord] = consumer.poll(poll_timeout_ms).get(topic_partition, [])
-                    if len(records) == 0:
-                        raise StaleConsumerError(topic_partition, start_offset, end_offset, last_offset, poll_timeout)
-                    for record in records:
-                        backend.store_record(fp, record)
-                        record_count += 1
-
-                    last_offset = record.offset  # pylint: disable=undefined-loop-variable
-                    if last_offset >= end_offset:
-                        break
-
-            destination = "stdout" if fp is sys.stdout else self.backup_location
-            print(
-                (
-                    f"Finished backup in format {version.name} of {topic}:{partition} "
-                    f"to {destination!r} (backed up {record_count:,} records)."
-                ),
-                file=sys.stderr,
+            except EmptyPartition:
+                LOG.warning(
+                    "Topic partition '%s' is empty, nothing to back up.",
+                    topic_partition,
+                )
+                return
+
+        end_time = datetime.datetime.now(datetime.timezone.utc)
+        if isinstance(self.backup_location, Path):
+            backend.store_metadata(
+                path=self.backup_location,
+                topic_name=self.topic_name,
+                topic_id=None,
+                started_at=start_time,
+                finished_at=end_time,
+                data_files=[data_file],
             )
+        else:
+            assert version is not BackupVersion.V3
```

Comment:
For pre-V3 only the calls to `backend.store_record()` in `_write_partition()` actually writes to files on disk, since we're not storing any topic-level metadata there. I'm not sure why I made the decision to make this backend call conditional though, since the default implementation is a no-op. I will most likely address this while implementing the wrapping directory discussed here: https://github.com/aiven/karapace/pull/584#discussion_r1198894936

I.e., I'll remove the if-else and always call `backend.store_metadata()` here.